### DISCUSSION TITLE
Help with marked point processes

### DISCUSSION THREAD

**Discussion by lcgodoy** [inlabru-org/inlabru#248] [Q&A] [answered]
Hello everyone,

## Summary

I am working with a marked point process such that the marks represent three categories. My goal is to estimate a surface representing the intensity of each of those processes (it is not realistic to assume they are independent).

I have tried three models so far. Model 1 is the simplest and comprises a log-Gaussian Cox process (LGCP) with a random intercept for each mark and a shared spatial effect. The second model considers a common intercept but a different spatial effect for each mark. Lastly, I considered a model with a common intercept, one common spatial effect, and a separate spatial effect for each mark.

This [tar file](https://github.com/user-attachments/files/16617519/inlabru-reprex.tar.gz) contains the data and the code to reproduce this example.

## Data and mesh

The data I am using is a `sf` object named `pts_m`. The columns that are important here are `mark`, `marks_int` and `geometry`. The map where the point process is observed is an `sfc` object called `boundary`.

## Model 1

I define this model as follows:

```
fm_1 <- geometry + marks_int ~ 0 +
  Marks_Intercept(marks_int, model = "iid",
                  hyper =
                    list(prec =
                           list(prior = "pc.prec",
                                param = c(0.01, 0.01)))) +
  field(geometry, model = matern)

strategy_inla <- "simplified.laplace" # default
int_strategy  <- "eb"

mod_1 <-
  lgcp(components = fm_1,
       data = pts_m,
       domain = list(geometry = mesh_fine,
                     marks_int = seq_len(max(pts_m$marks_int))),
       options = list(control.compute =
                        list(cpo = TRUE,
                             return.marginals.predictor = TRUE),
                      control.inla =
                        list(strategy = strategy_inla,
                             int.strategy = int_strategy),
                      verbose = FALSE,
                      safe = TRUE))

mod_1 <- bru_rerun(mod_1)
```

> I just realized that, for this model, I accidentally defined the “components” part of the model as having the “response” as 
> `geometry + marks_int`. How does this impact the model?

I get the following warning:
```
*** warning *** iterative process seems to diverge, 'vb.correction' is aborted
*** Please (re-)consider your model, priors, confounding, etc.
```
When I re-run (`bru_rerun`) the model, I get many of the following warnings : `GMRFLib_2order_approx: rescue NAN/INF values in logl for idx=18823`

## Model 2

Defined as:
```
fm_2 <- geometry ~ 0 + Intercept(1) +
  field(geometry, model = matern, group = marks_int)

mod_2 <-
  lgcp(components = fm_2,
       data = pts_m,
       domain = list(geometry = mesh_fine),
       options = list(control.compute =
                        list(cpo = TRUE,
                             return.marginals.predictor = TRUE),
                      control.inla =
                        list(strategy = strategy_inla,
                             int.strategy = int_strategy),
                      verbose = FALSE,
                      safe = TRUE))
```

No warnings are displayed. 

## Model 3

Code:

```
fm_3 <- geometry ~ 0 +
  Intercept(1) +
  common_field(geometry, model = matern) +
  marks_field(geometry,
              model = matern_int,
              group = marks_int)

mod_3 <-
  lgcp(components = fm_3,
       data = pts_m,
       ## domain = list(geometry = mesh_fine,
       ##               marks_int = seq_len(max(pts_m$marks_int))),
       domain = list(geometry = mesh_fine),
       options = list(control.compute =
                        list(cpo = TRUE,
                             return.marginals.predictor = TRUE),
                      control.inla =
                        list(strategy = strategy_inla,
                             int.strategy = int_strategy),
                      verbose = FALSE,
                      safe = TRUE))

mod_3 <- bru_rerun(mod_3)
```

I get many of the following warnings: `GMRFLib_2order_approx: rescue NAN/INF values in logl for idx=6275`
When re-running Model 3, I get only one of those warnings (as opposed to many). But I also get a warning about the divergence of the VB optimization (same as in the first model). 

## Problems/Questions

Any hints on workarounds here? I can think of three routes: (1) try tweaking the priors; (2) try to make the model simpler (probably not helpful since I need to predict the different surfaces for each mark); and (3) make the model more complex (perhaps non-stationarity of the Matern field? or the Barrier model?).

I am aware that another option is to work with [multiple likelihoods](https://inlabru-org.github.io/inlabru/articles/2d_lgcp_multilikelihood.html), but I could not make it work. This is probably because I have more than two marks, and I could not extend the example presented in the link.

Also, I am not sure if this approach I am taking is the most appropriate for marked-point processes. I would appreciate any suggestions.

### Model comparison

In terms of model comparison, @finnlindgren mentioned that WAIC is not defined for point processes. Is the LPML defined? 

### Predicted surface

When I try to predict the surfaces using the code below:
```
prediction_grid <-
  fm_pixels(mesh_fine,
            mask = boundary) |>
  mutate(mark_1 = 1,
         mark_2 = 2,
         mark_3 = 3) |>
  tidyr::pivot_longer(cols = c(mark_1:mark_3),
                      names_to = "garbage",
                      values_to = "marks_int") |>
  mutate(marks_int = as.integer(marks_int)) |>
  select(- garbage)

##--- * Model 1 ----

pred_1 <-
  predict(
      mod_1,
      prediction_grid,
      ~ data.frame(lambda = exp(Marks_Intercept + field))
  )

pred_1 <- left_join(pred_1,
                    y = {select(pts_m, mark, marks_int) |>
                           st_drop_geometry() |>
                           distinct()},
                    by = "marks_int”)

##--- * Model 2 ----

## matrix not (numerical) positive definite
pred_2 <-
  predict(
      mod_2,
      prediction_grid,
      ~ data.frame(lambda = exp(Intercept + field))
  )

##--- * Model 3 ----

## matrix not (numerical) positive definite
pred_3 <-
  predict(
      mod_3,
      prediction_grid,
      ~ data.frame(lambda = exp(Intercept + common_field + marks_field))
  )
```

I get errors saying the covariance matrix is not (numerical) positive definite for models 2 and 3. I assume that is common when there is a strong spatial dependence in the data, but I would appreciate any suggestion on how to deal with this in an applied setting.

### $K_{1, 2}$ functions

One last think I was wondering is if it is possible to get uncertainty quantifications on estimates of pairwise $K_{1, 2}$ functions once we manage to make the model converge. That would be interesting because it would allow to get an idea about repulsion or attraction relationships between the different processes.

Thanks a lot,
Lucas.

#### R session:

```
R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Los_Angeles
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base    

other attached packages:
 [1] dplyr_1.1.4         stars_0.6-6         abind_1.4-5        
 [4] INLAspacetime_0.1.8 ggplot2_3.5.1       inlabru_2.11.1    
 [7] fmesher_0.1.7       INLA_24.05.10       sp_2.1-4          
[10] Matrix_1.7-0        sf_1.0-16          

loaded via a namespace (and not attached):
 [1] s2_1.1.6            utf8_1.2.4          generics_0.1.3    
 [4] tidyr_1.3.1         class_7.3-22        KernSmooth_2.23-24
 [7] lattice_0.22-6      magrittr_2.0.3      grid_4.4.1        
[10] plyr_1.8.9          e1071_1.7-14        DBI_1.2.3          
[13] purrr_1.0.2         fansi_1.0.6         viridisLite_0.4.2  
[16] scales_1.3.0        numDeriv_2016.8-1.1 mnormt_2.1.1      
[19] cli_3.6.3           rlang_1.1.4         units_0.8-5        
[22] munsell_0.5.1       splines_4.4.1       withr_3.0.0        
[25] tools_4.4.1         parallel_4.4.1      MatrixModels_0.5-3
[28] colorspace_2.1-0    vctrs_0.6.5         R6_2.5.1          
[31] stats4_4.4.1        proxy_0.4-27        lifecycle_1.0.4    
[34] classInt_0.4-10     pkgconfig_2.0.3     pillar_1.9.0      
[37] gtable_0.3.5        glue_1.7.0          Rcpp_1.0.12        
[40] tibble_3.2.1        tidyselect_1.2.1    farver_2.1.2      
[43] labeling_0.4.3      wk_0.9.1            compiler_4.4.1    
[46] sn_2.1.1
```

**Comment by finnlindgren** [comment, marked as answer]
Hi,
there are a lot of inter-related questions here; it may be easier to break it down to some simpler sub-questions, and only move on from that once the issues with each step have been resolved.

Starting with model assessment, according to Aki Vehtari, LPML and WAIC are the same criteria: https://discourse.mc-stan.org/t/model-comparison-using-lpml-and-waic/27322/2
The issue is that they are based on "leave-one observation out" constructions, that don't make sense in the point process context. In the continuous domain point process context, one would have to leave out _a portion of space_; _space_ is the observation (or rather the information about where the observation method has attempted to detect points, and the location of those detected points; this is very different from the "value observed at a given location" context in which WAIC makes sense).

Second, one needs to be clear about what the model definition is, away from any details of how to express that as code. The fundamental defining property of a Poisson point process is that _by definition_, the distribution of the number of points contained in any given region, say $B$, takes a Poisson distribution where the expectation is the integral over the outcome space of an intensity function, $\lambda(s)$. If that expectation involves _anything_ that cannot be written as a pure function of $s$, the outcome space needs to be expanded to a combined space. For example, $\lambda(s)$ can depend on purely spatial covariates, such as elevation. However, it can _not_ depend on categorical variables associated with the points themselves. This means that e.g. species-specific $\lambda(s)$ functions require the outcome space to be augmented, for example by defining $\lambda(s, species)$, allowing species-specific point-intensities.  To _implement_ this, one would need `formula = geometry + species ~ ...` and `domain = list(geometry = mesh, species = species)` for the predictor and point process domain definitions. In your Model 3 definition, you omitted this explicit species dependence in the observation model, and only included it in the component definitions, but that violates the condition that $\lambda$ may depend _only_ on the quantities in the observation domain definition.

The issues with `Inf` messages etc cannot really be addressed until this fundamental model definition problem is corrected.

Note: the joint process construction discussed above is only required in the most general $\lambda$ structure. If the non-spatial observations (as you call "marks", indicating they are properties only defined for points, and not properties that exist in the absence of points) have distribution conditionally independent of location, one can write $\lambda(s,species) = \lambda(s) f(species)$, where the purely spatial intensity $\lambda$ controls the location distribution of the points, and $f(species)$ is a conditionally independent species allocation model. In this case, one can write a purely spatial point process model based on space alone, and a separate `like()` model for the marks. However, this doesn't work if that distribution needs to depend on location, with $f(species|s)$. In such cases, the full $\lambda(s,species)$ augmented observation space approach is needed, e.g. with predictor $\log(\lambda(s,species)) = \log(\lambda(s)) + \log(f(species|s))$ or the more "free-form" approach your model 3 component definition implies, with a direct model for $\log(\lambda(s,species))$.

**Reply by finnlindgren** [reply]
Correction: I think having location dependent $f(species|s)$ is ok also for using a separate `like()` for the marks. The important issue is if $\lambda(s,species)$ can be factorised into $\lambda(s) f(species|s)$ or not. A case where this factorisation cannot be done is when the detectability of a point depends on the mark or if there are common effect in the $\lambda$ and $f$ models; e.g. if large birds are easier to detect than small birds. In such cases, $\lambda(s)$ needs to involve the mark directly, and the factorisation isn't possible.

Another case that is easiest to implement as a $\lambda(s,species)$ model is when one models the intensity field for each species, but with potential common effects between the species.

**Comment by lcgodoy** [comment]
Hi Finn,

Thank you so much for your thorough answer. I don’t think the factoriation can be done, as the detection probabilities are different across different marks. I am going to think harder about the assumptions and how to formalize the model. I will reach out if the problems when fitting the model persist.

Thanks a lot again,
Lucas.

**Reply by finnlindgren** [reply]
Ah, yes, mark-dependent detection probabilities need the joint lambda(s, mark) formulation.
